File: Crio/crio_programming_interview_problems-master/PatternMatching/PatternMatching.java
import java.util.*;

public class PatternMatching {

    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    String text , pattern;
    boolean answer;
    Map<Character, String> map;
    void isMatch(int i , int j) {

        if (i == this.text.length()) {
            if (j == this.pattern.length()) {
                this.answer = true;
            }
            return;
        }
        if (j == this.pattern.length()) {
            return;
        }
        char patternPart = this.pattern.charAt(j);
        if (this.map.containsKey(patternPart)) {
            String textPart = this.map.get(patternPart);
            int length = textPart.length();
            if (length > text.length() - i) {
                return;
            } else {
                String subsString = text.substring(i, i + length);
                if (subsString.equals(textPart)) {
                    isMatch(i + length , j + 1);
                } else {
                    return;
                }
            }
        } else {
            String value = "";
            for (int k = 1 ; i + k - 1 < this.text.length() ; k++) {
                value += this.text.charAt(i + k - 1);
                this.map.put(pattern.charAt(j), value);
                this.isMatch(i + k, j + 1);
                this.map.remove(pattern.charAt(j));
            }
        }

    }
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

    boolean patternMatching(String text, String pattern) {

        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        this.answer = false;
        this.map = new HashMap<Character, String>();
        this.text = text;
        this.pattern = pattern;
        isMatch(0, 0);
        return answer;
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        for (int p = 0 ; p < t ; p++) {
            PatternMatching result = new PatternMatching();
            String text = scanner.next();
            String pattern = scanner.next();
            if (result.patternMatching(text, pattern)) {
                System.out.println("Yes");
            } else {
                System.out.println("No");
            }
        }
        scanner.close();
    }
}

File: Crio/crio_programming_interview_problems-master/PatternMatching/PatternMatching.py
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
def isMatch(i, j, prevI, wordMap, text, pattern):
    n = len(text)
    m = len(pattern)
    
    if i == n and j == m:
        return True
    if i == n or j == m:
        return False
    
    result = isMatch(i + 1, j, prevI, wordMap, text, pattern)

    currentStr = text[prevI + 1: i + 1]

    if pattern[j] not in wordMap:
        wordMap[pattern[j]] = currentStr
        result = result or isMatch(i + 1, j + 1, i, wordMap, text, pattern)
        wordMap.pop(pattern[j])
    else:
        if currentStr == wordMap[pattern[j]]:
            result = result or isMatch(i + 1, j + 1, i, wordMap, text, pattern)
    return result
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def patternMatching(text, pat):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    return isMatch(0, 0, -1, {}, text, pattern)
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

T = int(input())
for _t in range(T):
    text, pattern = input().strip().split()
    result = patternMatching(text, pattern)
    print("Yes" if result else "No")
